src/System.IO.FileSystem/src/System/IO/DosMatcher.cs

```diff
@@ -0,0 +1,307 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Text;
+
+namespace System.IO
+{
+    internal static class DosMatcher
+    {
+        // [MS - FSA] 2.1.4.4 Algorithm for Determining if a FileName Is in an Expression
+        // https://msdn.microsoft.com/en-us/library/ff469270.aspx
+        private static readonly char[] s_wildcardChars =
+        {
+            '\"', '<', '>', '*', '?'
+        };
+
+        /// <summary>
+        /// Change '*' and '?' to '&lt;', '&gt;' and '"' to match Win32 behavior. For compatibility, Windows
+        /// changes some wildcards to provide a closer match to historical DOS 8.3 filename matching.
+        /// </summary>
+        internal static string TranslateExpression(string expression)
+        {
+            if (string.IsNullOrEmpty(expression) || expression == "*.*")
+                return "*";
+
+            StringBuilder sb = StringBuilderCache.Acquire(expression.Length);
+            int length = expression.Length;
+            for (int i = 0; i < length; i++)
+            {
+                char c = expression[i];
+                switch (c)
+                {
+                    case '.':
+                        if (i > 1 && i == length - 1 && expression[i - 1] == '*')
+                        {
+                            sb.Length--;
+                            sb.Append('<'); // DOS_STAR (ends in *.)
+                        }
+                        else if (i < length - 1 && (expression[i + 1] == '?' || expression[i + 1] == '*'))
+                        {
+                            sb.Append('\"'); // DOS_DOT
+                        }
+                        else
+                        {
+                            sb.Append('.');
+                        }
+                        break;
+                    case '?':
+                        sb.Append('>'); // DOS_QM
+                        break;
```

Comment:
Did you consider tracking whether you ever hit the '.' or '?' cases?  If you never do, you can just return the original string rather than creating a new one, right?

Reply:
Didn't, but I will.